api/submission/remove_submission_test.go

```diff
@@ -0,0 +1,95 @@
+package submission
+
+import (
+    "testing"
+
+    "github.com/eriq-augustine/autograder/api/core"
+    "github.com/eriq-augustine/autograder/db"
+    "github.com/eriq-augustine/autograder/model"
+    "github.com/eriq-augustine/autograder/util"
+)
+
+func TestRemoveSubmission(test *testing.T) {
+    // Leave the course in a good state after the test.
+    defer db.ResetForTesting();
+
+    testCases := []struct{ role model.UserRole; targetEmail string; targetSubmission string; foundUser bool; foundSubmission bool; permError bool}{
+        // Grader, self, recent.
+        {model.RoleGrader, "",                "", true, false, false},
+        {model.RoleGrader, "grader@test.com", "", true, false, false},
+
+        // Grader, self, missing.
+        {model.RoleGrader, "",                "ZZZ", true, false, false},
+        {model.RoleGrader, "grader@test.com", "ZZZ", true, false, false},
+
+        // Grader, other, recent.
+        {model.RoleGrader, "student@test.com", "", true, true, false},
+
+        // Grader, other, specific.
+        {model.RoleGrader, "student@test.com", "1697406256", true, true, false},
+        {model.RoleGrader, "student@test.com", "1697406265", true, true, false},
+        {model.RoleGrader, "student@test.com", "1697406272", true, true, false},
+
+        // Grader, other, specific (full ID).
+        {model.RoleGrader, "student@test.com", "course101::hw0::student@test.com::1697406256", true, true, false},
+        {model.RoleGrader, "student@test.com", "course101::hw0::student@test.com::1697406265", true, true, false},
+        {model.RoleGrader, "student@test.com", "course101::hw0::student@test.com::1697406272", true, true, false},
+
+        // Grader, other, missing.
+        {model.RoleGrader, "student@test.com", "ZZZ", true, false, false},
+
+        // Grader, missing, recent.
+        {model.RoleGrader, "ZZZ@test.com", "", false, false, false},
+
+        // Roles below grader, other, recent.
+        {model.RoleStudent, "student@test.com", "", false, false, true},
+        {model.RoleOther,   "student@test.com", "", false, false, true},
+        {model.RoleUnknown, "student@test.com", "", false, false, true},
+
+        // Roles above grader, other, recent
+        {model.RoleAdmin, "student@test.com", "", true, true, false},
+        {model.RoleOwner, "student@test.com", "", true, true, false},
+    };
+
+    for i, testCase := range testCases {
+        // Reload the test course every time.
+        db.ResetForTesting();
+
+        fields := map[string]any{
+            "target-email": testCase.targetEmail,
+            "target-submission": testCase.targetSubmission,
+        };
+
+        response := core.SendTestAPIRequestFull(test, core.NewEndpoint(`submission/remove/submission`), fields, nil, testCase.role);
+        
+        if (!response.Success) {
+            if (testCase.role == model.RoleUnknown) {
```

Comment:
Ok, I guess I'm just wondering when to draw the line between when to add tests to be extra safe vs when tests are not necessary.

Reply:
That can be a hard thing to figure out, and it can depend on the organization, project, and timeline.

But in this case it is a little different.
Since RoleUnknown is different than all the other roles, it does not need to be tested like the other roles.
It is handled at a much earlier stage (in the API's core), so anyone further down (like an endpoint) doesn't need to worry about it.